drop-transfer/src/file/mod.rs

```diff
@@ -202,58 +221,59 @@ impl File {
-        &self.file_id
-    }
+        let mut files = Vec::new();
+        let mut breadth = 0;
 
-    pub fn subpath(&self) -> &FileSubPath {
-        &self.subpath
-    }
+        for entry in WalkDir::new(path).min_depth(1).into_iter() {
+            let entry = entry?;
+            let meta = entry.metadata()?;
+
+            if !meta.is_file() {
+                continue;
+            }
 
-    pub fn info(&self) -> Option<FileInfo> {
-        match &self.kind {
-            FileKind::FileToSend { mime_type, .. } => Some(FileInfo {
-                mime_type: mime_type
-                    .as_deref()
-                    .cloned()
-                    .unwrap_or_else(|| "unknown".to_string()),
-                extension: AsRef::<Path>::as_ref(self.subpath.name())
-                    .extension()
-                    .map(|e| e.to_string_lossy().to_string())
-                    .unwrap_or_else(|| "unknown".to_string()),
-                size_kb: (self.size() as f64 / 1024.0).ceil() as i32,
-            }),
-            _ => None,
+            if entry.depth() > config.dir_depth_limit {
+                return Err(Error::TransferLimitsExceeded);
+            }
+
+            breadth += 1;
+
+            if breadth > config.transfer_file_limit {
+                return Err(Error::TransferLimitsExceeded);
+            }
+
+            let subpath = entry
+                .path()
+                .strip_prefix(parent)
+                .map_err(|err| crate::Error::BadPath(err.to_string()))?;
+
+            let subpath = FileSubPath::from_path(subpath)?;
+
+            let path = entry.into_path();
+            let file_id = file_id_from_path(&path)?;
+            let file = Self::new_to_send(subpath, path, meta, file_id)?;
+            files.push(file);
         }
+
+        Ok(files)
     }
 
     // Open the file if it wasn't already opened and return the std::fs::File
     // instance
     pub(crate) fn open(&self, offset: u64) -> crate::Result<FileReader> {
-        match &self.kind {
-            FileKind::FileToSend { meta, source, .. } => {
-                let mut reader = reader::open(source)?;
-                reader.seek(io::SeekFrom::Start(offset))?;
-                FileReader::new(reader, meta.0.clone())
-            }
-            _ => Err(Error::BadFile),
-        }
+        let mut reader = reader::open(&self.source)?;
+        reader.seek(io::SeekFrom::Start(offset))?;
+        FileReader::new(reader, self.meta.0.clone())
     }
 
     /// Calculate sha2 of a file. This is a blocking operation
     pub(crate) fn checksum(&self, limit: u64) -> crate::Result<[u8; 32]> {
-        let reader = match &self.kind {
-            FileKind::FileToSend { source, .. } => reader::open(source)?,
-            _ => return Err(Error::BadFile),
-        };
-
+        let reader = reader::open(&self.source)?;
         let mut reader = io::BufReader::new(reader).take(limit);
         let csum = checksum(&mut reader)?;
```

Comment:
Can we be sure BadFile is still being returned in such a case?

Reply:
`BadFile` was returned when trying to open a file to be received. Now with strong typing, there is no way one could open a wrong file, so the error is gone

Reply:
Nice!